netlink: network namespaces can be disabled in kernel configuration

When I try to use the library in order to open a generic netlink socket I get the error message:
```
panic: failed to dial: open /proc/self/task/2281/ns/net: no such file or directory
```

It seems that this library expects a feature to be enabled in my kernel which is not. Do you happen to know which kernel build configuration/feature is responsible for this subdirectory? It might be handy to document this requirement as well.

--- a/conn_linux.go
+++ b/conn_linux.go
@@ -338,7 +338,8 @@ type sysSocket struct {
 // newSysSocket creates a sysSocket that optionally locks its internal goroutine
 // to a single thread.
 func newSysSocket(config *Config) (*sysSocket, error) {
-	g, err := newLockedNetNSGoroutine(config.NetNS)
+	// Determine network namespaces using the threadNetNS function.
+	g, err := newLockedNetNSGoroutine(config.NetNS, threadNetNS)
 	if err != nil {
 		return nil, err
 	}
@@ -620,11 +621,14 @@ type lockedNetNSGoroutine struct {
 	funcC chan func()
 }
 
-func newLockedNetNSGoroutine(netNS int) (*lockedNetNSGoroutine, error) {
+// newLockedNetNSGoroutine creates a lockedNetNSGoroutine that will enter the
+// specified network namespace netNS (by file descriptor), and will use the
+// getNS function to produce netNS handles.
+func newLockedNetNSGoroutine(netNS int, getNS func() (*netNS, error)) (*lockedNetNSGoroutine, error) {
 	// Any bare syscall errors (e.g. setns) should be wrapped with
 	// os.NewSyscallError for the remainder of this function.
 
-	callerNS, err := threadNetNS()
+	callerNS, err := getNS()
 	if err != nil {
 		return nil, err
 	}
@@ -662,7 +666,7 @@ func newLockedNetNSGoroutine(netNS int) (*lockedNetNSGoroutine, error) {
 		defer g.wg.Done()
 
 		// Get the current namespace of the thread the goroutine is locked to.
-		threadNS, err := threadNetNS()
+		threadNS, err := getNS()
 		if err != nil {
 			errC <- err
 			return

--- a/conn_linux_test.go
+++ b/conn_linux_test.go
@@ -500,6 +500,49 @@ func TestLinuxConnConfig(t *testing.T) {
 	}
 }
 
+func Test_newLockedNetNSGoroutineNetNSDisabled(t *testing.T) {
+	tests := []struct {
+		name string
+		ns   int
+		ok   bool
+	}{
+		{
+			// Network namespaces are disabled but none is set: this should
+			// succeed.
+			name: "not set",
+			ok:   true,
+		},
+		{
+			// Network namespaces are disabled but one is set explicitly:
+			// this should fail.
+			name: "set",
+			ns:   1,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g, err := newLockedNetNSGoroutine(tt.ns, func() (*netNS, error) {
+				// Network namespaces should be disabled due to a non-existent
+				// file.
+				return fileNetNS("/netlinktestdoesnotexist")
+			})
+			if err != nil {
+				if tt.ok {
+					t.Fatalf("failed to create goroutine: %v", err)
+				}
+
+				return
+			}
+			defer g.stop()
+
+			if !tt.ok {
+				t.Fatal("expected an error, but none occurred")
+			}
+		})
+	}
+}
+
 func testLinuxConn(t *testing.T, config *Config) (*conn, *testSocket) {
 	s := &testSocket{}
 	c, _, err := bind(s, config)

--- a/netns_linux.go
+++ b/netns_linux.go
@@ -31,7 +31,13 @@ type netNS struct {
 // thread. If the namespace is not the default namespace, runtime.LockOSThread
 // should be invoked first.
 func threadNetNS() (*netNS, error) {
-	f, err := os.Open(fmt.Sprintf("/proc/self/task/%d/ns/net", unix.Gettid()))
+	return fileNetNS(fmt.Sprintf("/proc/self/task/%d/ns/net", unix.Gettid()))
+}
+
+// fileNetNS opens file and creates a netNS. fileNetNS should only be called
+// directly in tests.
+func fileNetNS(file string) (*netNS, error) {
+	f, err := os.Open(file)
 	switch {
 	case err == nil:
 		return &netNS{f: f}, nil

--- a/netns_linux.go
+++ b/netns_linux.go
@@ -3,18 +3,28 @@
 package netlink
 
 import (
+	"errors"
 	"fmt"
 	"os"
 
 	"golang.org/x/sys/unix"
 )
 
+// errNetNSDisabled is returned when network namespaces are unavailable on
+// a given system.
+var errNetNSDisabled = errors.New("netlink: network namespaces are not enabled on this system")
+
 // A netNS is a handle that can manipulate network namespaces.
 //
 // Operations performed on a netNS must use runtime.LockOSThread before
 // manipulating any network namespaces.
 type netNS struct {
+	// The handle to a network namespace.
 	f *os.File
+
+	// Indicates if network namespaces are disabled on this system, and thus
+	// operations should become a no-op or return errors.
+	disabled bool
 }
 
 // threadNetNS constructs a netNS using the network namespace of the calling
@@ -22,23 +32,56 @@ type netNS struct {
 // should be invoked first.
 func threadNetNS() (*netNS, error) {
 	f, err := os.Open(fmt.Sprintf("/proc/self/task/%d/ns/net", unix.Gettid()))
-	if err != nil {
+	switch {
+	case err == nil:
+		return &netNS{f: f}, nil
+	case os.IsNotExist(err):
+		// Network namespaces are not enabled on this system. Use this signal
+		// to return errors elsewhere if the caller explicitly asks for a
+		// network namespace to be set.
+		return &netNS{disabled: true}, nil
+	default:
 		return nil, err
 	}
-
-	return &netNS{f: f}, nil
 }
 
 // Close releases the handle to a network namespace.
-func (n *netNS) Close() error { return n.f.Close() }
+func (n *netNS) Close() error {
+	return n.do(func() error {
+		return n.f.Close()
+	})
+}
 
 // FD returns a file descriptor which represents the network namespace.
-func (n *netNS) FD() int { return int(n.f.Fd()) }
+func (n *netNS) FD() int {
+	if n.disabled {
+		// No reasonable file descriptor value in this case, so specify a
+		// non-existent one.
+		return -1
+	}
+
+	return int(n.f.Fd())
+}
 
 // Restore restores the original network namespace for the calling thread.
-func (n *netNS) Restore() error { return n.Set(n.FD()) }
+func (n *netNS) Restore() error {
+	return n.do(func() error {
+		return n.Set(n.FD())
+	})
+}
 
 // Set sets a new network namespace for the current thread using fd.
 func (n *netNS) Set(fd int) error {
-	return os.NewSyscallError("setns", unix.Setns(fd, unix.CLONE_NEWNET))
+	return n.do(func() error {
+		return os.NewSyscallError("setns", unix.Setns(fd, unix.CLONE_NEWNET))
+	})
+}
+
+// do runs fn if network namespaces are enabled on this system.
+func (n *netNS) do(fn func() error) error {
+	if n.disabled {
+		return errNetNSDisabled
+	}
+
+	return fn()
 }

